import "ydb/core/protos/base.proto";
import "ydb/core/scheme/protos/type_info.proto";

package NKikimr.NTable.NProto;
option java_package = "ru.yandex.kikimr.proto";

message TEvol {
    required uint32 Tail  = 1;
    required uint32 Head  = 2;
}

message TPartColumn {
    optional uint32 Tag = 1;
    optional uint32 Type = 2;
    optional NKikimrProto.TTypeInfo TypeInfo = 5;
    optional uint32 Key = 3;
    optional uint32 Group = 4;
}

message TPartScheme {
    repeated TPartColumn Columns = 1;
    repeated uint32 KeyTags = 2;
    optional uint32 Legacy0_ = 3;
}

message TBTreeIndexMeta {
    optional uint32 RootPageId = 1;
    optional uint32 LevelCount = 2;
    optional uint64 IndexSize = 3;
    optional uint64 DataSize = 4; // Main pages data size
    optional uint64 RowCount = 5; // Total number of data rows
    optional uint64 ErasedRowCount = 6; // Total number of erased data rows
    optional uint64 GroupDataSize = 7; // Group pages data size (includes history and external blobs)
}

message TLayout {
    optional uint32 Index = 1;
    optional uint32 Globs = 2;  // EPage::Globs, external blobs catalog
    optional uint32 Large = 3;  // EPage::Frames index for globs catalog
    optional uint32 Small = 6;  // EPage::Frames index for packed blobs
    optional uint32 ByKey = 4;  // EPage::ByKey bloom filter page
    optional uint32 Scheme = 5; // EPage::Schem2 with data pages scheme
    repeated uint32 GroupIndexes = 7; // EPage::FlatIndex for alt column groups
    repeated uint32 HistoricIndexes = 8; // EPage::FlatIndex for historic data
    optional uint32 GarbageStats = 9; // EPage::GarbageStats for historic data
    optional uint32 TxIdStats = 10; // EPage::TxIdStats for uncommitted deltas
    repeated TBTreeIndexMeta BTreeGroupIndexes = 11;
    repeated TBTreeIndexMeta BTreeHistoricIndexes = 12;
    optional uint32 BTreeIndexesFormatVersion = 13;
}

message TStat {
    optional uint64 Bytes = 1;  // Acrual raw data pages size in part
    optional uint64 Coded = 2;  // Observable (encoded) data pages size
    optional uint64 Drops = 3;  // Rows with ERowOp::Erase code
    optional uint64 Rows = 4;   // All rows count in the TPart
    optional uint64 HiddenRows = 5; // Hidden (non-head) total rows
    optional uint64 HiddenDrops = 6; // Hidden (non-main) rows with ERowOp::Erase
}

message TRoot { // NTable::TPart page collection metablob
    optional int64 Epoch = 2;
    optional uint64 Bytes = 3;  // Actual raw data pages size in part
    optional uint64 Coded = 4;  // Observable (encoded) data pages size
    optional TEvol Evol = 6;
    optional TStat Stat = 7;   // Some stats collected on compaction
    optional TLayout Layout = 8;// Non-data pages layout information
    optional NKikimrProto.TRowVersion MinRowVersion = 9;
    optional NKikimrProto.TRowVersion MaxRowVersion = 10;
}

message TSliceBounds {
    optional bytes FirstKey = 1;
    optional bytes LastKey = 2;
    optional uint64 FirstRowId = 3;
    optional uint64 LastRowId = 4;
    optional bool FirstInclusive = 5;
    optional bool LastInclusive = 6;
    reserved 7; // uint64 DataSize no longer used or stored
}

message TOverlay {
    repeated uint64 Screen = 1;     // Set of [begin, end) rowId ranges
    repeated TSliceBounds Slices = 2; // Set of extended range slices
}

message TOverlayDelta {
    repeated TSliceBounds RemovedSlices = 1; // set of slices that are removed
    repeated TSliceBounds ChangedSlices = 2; // set of slices that replace identical row ranges
}
